### PR TITLE
wolfcrypt: support NIST 800-56C Option 1 KDF

### PR DESCRIPTION
**Pull request by rizlik**
# Description

 support NIST 800-56C Option 1 KDF

Fixes zd#17977

# Testing

- Added test with NIST provided test vectors

# Checklist

 - [X] added tests
 - [X] updated/added doxygen
 - [X] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/kdf.c [hunk 1]
```diff
@@ -1389,4 +1389,123 @@ int wc_SRTP_KDF_kdr_to_idx(word32 kdr)
...
+ * \param [in] derivedSecretSz The desired size of the derived secret.
+ * \param [in] hashType The hash algorithm to be used in the KDF.
+ * \param [out] output The buffer to store the derived secret.
+ * \param [in] outputSz The size of the output buffer.
+ *
+ * \return 0 if the KDF operation is successful.
+ * \return BAD_FUNC_ARG if the input parameters are invalid.
+ * \return negative error code if the KDF operation fails.
+ */
+int wc_SP80056C_KDF_single(const byte* z, word32 zSz,
```
Comment (SparkiDev) [inline wolfcrypt/src/kdf.c:1436]: Better name is wc_KDA_KDF_onestep().
KDA appears to be the name they are giving the algorithms.
KDF as it is the hash version. (HKDF for HMAC)
onestep as that's how they describe it.
https://pages.nist.gov/ACVP/draft-hammett-acvp-kas-kdf-onestep.html

File: wolfcrypt/src/kdf.c [hunk 2]
```diff
@@ -1389,4 +1389,123 @@ int wc_SRTP_KDF_kdr_to_idx(word32 kdr)
...
+            return BAD_FUNC_ARG;
+        reps++;
+    }
+
+    /* According to SP800_56C, table 1, the max input size (max_H_inputBits)
+     * depends on the HASH algo. The smaller value in the table is (2**64-1)/8.
+     * This is larger than the possible length using word32 integers. The code
+     * checks for overflow. */
+    inputSz = zSz;
+    if (inputSz + 4 < inputSz)
```
Comment (SparkiDev) [inline wolfcrypt/src/kdf.c:1473]: This looks weird.
(a + 4) < a

File: wolfcrypt/src/kdf.c [hunk 3]
```diff
@@ -1389,4 +1389,123 @@ int wc_SRTP_KDF_kdr_to_idx(word32 kdr)
...
+         * Nevertheless, the code conservatively does the copy anyway as the
+         * data is sensitive and the user may forget zeroing outputsz bytes
+         * instead of derivedSecretsz bytes. */
+        if (outIdx + hashOutSz <= derivedSecretSz) {
+            hashOut = output + outIdx;
+        }
+        else {
+            hashOut = hashTempBuf;
+        }
+        ret = wc_SP80056C_KDF_iteration(z, zSz, counter,
```
Comment (SparkiDev) [inline wolfcrypt/src/kdf.c:1492]: Change name of this function to wc_KDA_KDF_iteration.

File: wolfcrypt/src/kdf.c [hunk 4]
```diff
@@ -1389,4 +1389,123 @@ int wc_SRTP_KDF_kdr_to_idx(word32 kdr)
...
+    if (derivedSecretSz == 0)
+        return BAD_FUNC_ARG;
+
+    hashOutSz = wc_HashGetDigestSize(hashType);
+    if (hashOutSz == HASH_TYPE_E)
+        return BAD_FUNC_ARG;
+
+    /* According to SP800_56C reps shall not be greater than 2**32-1. This is
+     * not possible using word32 integers. The code checks for overflow. */
+    reps = derivedSecretSz / hashOutSz;
```
Comment (SparkiDev) [inline wolfcrypt/src/kdf.c:1461]: reps = (derivedSecretSz + hashOutSz - 1) / hashOutSz

File: wolfcrypt/src/kdf.c [hunk 5]
```diff
@@ -1389,4 +1389,123 @@ int wc_SRTP_KDF_kdr_to_idx(word32 kdr)
...
+     * checks for overflow. */
+    inputSz = zSz;
+    if (inputSz + 4 < inputSz)
+        return BAD_FUNC_ARG;
+    inputSz += 4;
+    if (inputSz + fixedInfoSz < inputSz)
+        return BAD_FUNC_ARG;
+
+    outIdx = 0;
+    for (counter = 1; counter <= reps; counter++) {
```
Comment (SparkiDev) [inline wolfcrypt/src/kdf.c:1480]: Could break this into two.
One loop for whole and a single call that puts result into temp if there is less than a block needed at end.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/kdf.c (hunk 2 above)
Comment (rizlik) [reply wolfcrypt/src/kdf.c]: Removed the overflow checks: inputs are used one at a time inside wc_KDA_KDF_iteration(). Also, the sum of their size can't exceed the maximum specified inside SP800-56C ((2**64-1)/8)

File: wolfcrypt/src/kdf.c (hunk 4 above)
Comment (rizlik) [reply wolfcrypt/src/kdf.c]: removed as reps variable is not needed after refactoring

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**